sources/fitbit.py

```diff
@@ -0,0 +1,379 @@
+            today_date = arrow.get()
+
+            dates = []
+
+            for year in xrange(start_date.year, today_date.year + 1):
+                start_month = 1
+                end_month = 12
+
+                if year == start_date.year == today_date.year:
+                    start_month = start_date.month
+                    end_month = today_date.month
+                elif year == start_date.year:
+                    start_month = start_date.month
+                elif year == today_date.year:
+                    end_month = today_date.month
+
+                dates += [(year, month) for month
+                          in range(start_month, end_month + 1)]
+
+            for year, month in dates:
+                print 'retrieving {}: {}, {}'.format(url['name'], year, month)
+
+                fitbit_data[url['name']] = []
+
+                query_result = fitbit_query(
+                    access_token=access_token,
+                    path=url['url'],
+                    parameters={
+                        'user_id': user_id,
+                        'date': '{}-{:02d}-01'.format(year, month),
+                    })
+
+                fitbit_data[url['name']].append(query_result)
+
+        # TODO: implement these once we're approved for Fitbit intraday access
+        for url in [url for url in fitbit_urls if url['period'] == 'day']:
+            pass
+    except RateCapException:
+        return {
+            'all_data': fitbit_data,
+            'rate_cap_encountered': True,
+        }
+
+    return {
+        'all_data': fitbit_data,
+        'rate_cap_encountered': False,
+    }
+
+
+def create_datafiles(access_token, task_id=None, update_url=None, **kwargs):
+    """
+    Create Open Humans Dataset from Fitbit API data
+
+    Required arguments:
+        access_token: Fitbit access token
+
+    Optional arguments:
+        filedir: Local filepath, folder in which to place the resulting file.
+        s3_bucket_name: S3 bucket to write resulting file.
+        s3_key_dir: S3 key "directory" to write resulting file. The full S3 key
+                    name will add a filename to the end of s3_key_dir.
+
+    Either 'filedir' (and no S3 arguments), or both S3 arguments (and no
+    'filedir') must be specified.
+    """
+    tempdir = tempfile.mkdtemp()
+    filename = 'fitbit-data.json'
+    filepath = os.path.join(tempdir, filename)
+    data_files = []
+
+    temp_files = [{
+        'temp_filename': filename,
+        'tempdir': tempdir,
+        'metadata': {
+            'description': ('Fitbit activity, health, and fitness data.'),
+            'tags': ['weight', 'Fitbit', 'steps', 'activity'],
+        }
+    }]
+
+    fitbit_data = get_fitbit_data(access_token)
+
+    if fitbit_data['rate_cap_encountered']:
+        # If this is previously called and we got no new data this round,
+        # double the wait period for resubmission.
+        if 'return_status' in kwargs and not fitbit_data['all_data']:
```

Comment:
Confession: I think this is copied from my approach in Moves, and I'm pretty sure it isn't working like it should. Cached data is going to result in 'all_data' being filled, even without any successful API calls. (I don't think this is a blocker.)